gems/aws-sigv4/lib/aws-sigv4/signer.rb

```diff
@@ -0,0 +1,335 @@
+      #   `#credentials`, returning an object that responds to the following
+      #   methods:
+      #
+      #   * `#access_key_id` => String
+      #   * `#secret_access_key` => String
+      #   * `#session_token` => String, nil
+      #   * `#set?` => Boolean
+      #
+      attr_reader :credentials_provider
+
+      # @return [Set<String>] Returns a set of header names that should not be signed.
+      #   All header names have been downcased.
+      attr_reader :unsigned_headers
+
+      # @return [Boolean] When `true` the `x-amz-content-sha256` header will be signed and
+      #   returned in the signature headers.
+      attr_reader :apply_checksum_header
+
+      # Computes a version 4 signature signature. Returns the resultant
+      # signature as a hash of headers to apply to your HTTP request. The given
+      # request is not modified.
+      #
+      #     signature = signer.sign_request(
+      #       http_method: 'PUT',
+      #       url: 'https://domain.com',
+      #       headers: {
+      #         'Abc' => 'xyz',
+      #       },
+      #       body: 'body' # String or IO object
+      #     )
+      #
+      #     # Apply the following hash of headers to your HTTP request
+      #     signature.headers['host']
+      #     signature.headers['x-amz-date']
+      #     signature.headers['x-amz-security-token']
+      #     signature.headers['x-amz-content-sha256']
+      #     signature.headers['authorization']
+      #
+      # In addition to computing the signature headers, the canonicalized
+      # request, string to sign and content sha256 checksum are also available.
+      # These values are useful for debugging signature errors returned by AWS.
+      #
+      #     signature.canonical_request #=> "..."
+      #     signature.string_to_sign #=> "..."
+      #     signature.content_sha256 #=> "..."
+      #
+      # @param [Hash] request
+      #
+      # @option request [required, String] :http_method One of
+      #   'GET', 'HEAD', 'PUT', 'POST', 'PATCH', or 'DELETE'
+      #
+      # @option request [required, String, URI::HTTPS, URI::HTTP] :url
+      #   The request URI. Must be a valid HTTP or HTTPS URI.
+      #
+      # @option request [optional, Hash] :headers ({}) A hash of headers
+      #   to sign. If the 'X-Amz-Content-Sha256' header is set, the `:body`
+      #   is optional and will not be read.
+      #
+      # @option request [optional, String, IO] :body ('') The HTTP request body.
+      #   A sha256 checksum is computed of the body unless the
+      #   'X-Amz-Content-Sha256' header is set.
+      #
+      # @return [Signature] Return an instance of {Signature} that has
+      #   a `#headers` method. The headers must be applied to your request.
+      #
+      def sign_request(request)
+        creds = fetch_credentials
+
+        http_method = extract_http_method(request)
+        url = extract_url(request)
+        headers = downcase_headers(request[:headers])
```

Comment:
shouldn't need to dowcase the headers. aws-c-auth aught to do that

Reply:
Some of the headers returned on SigningResult are not downcased: I get Authorization  and X-Amz-Date back